kube-client/src/api/portforward.rs

```diff
@@ -0,0 +1,225 @@
+type ErrorSender = oneshot::Sender<String>;
+
+enum Message {
+    FromPod(Vec<u8>),
+    ToPod(Vec<u8>),
+}
+
+struct PortforwarderState {
+    waker: Option<Waker>,
+    finished: bool,
+    ports: Option<Vec<Port>>,
+}
+
+// Provides `AsyncRead + AsyncWrite` for each port and **does not** bind to local ports.
+// Error channel for each port is only written by the server when there's an exception and
+// the port cannot be used (didn't initialize or can't be used anymore).
+/// Manage port forwarding.
+pub struct Portforwarder {
+    state: Arc<Mutex<PortforwarderState>>,
+}
+
+impl Portforwarder {
+    pub(crate) fn new<S>(stream: WebSocketStream<S>, port_nums: &[u16]) -> Self
+    where
+        S: AsyncRead + AsyncWrite + Unpin + Sized + Send + 'static,
+    {
+        let mut ports = Vec::new();
+        let mut errors = Vec::new();
+        let mut duplexes = Vec::new();
+        for _ in port_nums.iter() {
+            let (a, b) = tokio::io::duplex(1024 * 1024);
+            let (tx, rx) = oneshot::channel();
+            ports.push(Port::new(a, rx));
+            errors.push(Some(tx));
+            duplexes.push(b);
+        }
+
+        let state = Arc::new(Mutex::new(PortforwarderState {
+            waker: None,
+            finished: false,
+            ports: Some(ports),
+        }));
+        let shared_state = state.clone();
+        let port_nums = port_nums.to_owned();
+        tokio::spawn(async move {
+            start_message_loop(stream, &port_nums, duplexes, errors).await;
+
+            let mut shared = shared_state.lock().unwrap();
+            shared.finished = true;
+            if let Some(waker) = shared.waker.take() {
+                waker.wake()
+            }
+        });
+        Portforwarder { state }
+    }
+
+    /// Get streams for forwarded ports.
+    pub fn ports(&mut self) -> Option<Vec<Port>> {
+        let mut state = self.state.lock().unwrap();
+        state.ports.take()
+    }
+}
+
+impl Future for Portforwarder {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut state = self.state.lock().unwrap();
+        if state.finished {
+            return Poll::Ready(());
+        }
+
+        if let Some(waker) = &state.waker {
+            if waker.will_wake(cx.waker()) {
+                return Poll::Pending;
+            }
+        }
+
+        state.waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+pub struct Port {
+    // Data pipe.
+    stream: Option<DuplexStream>,
+    // Error channel.
+    error: Option<ErrorReceiver>,
+}
+
+impl Port {
+    pub(crate) fn new(stream: DuplexStream, error: ErrorReceiver) -> Self {
+        Port {
+            stream: Some(stream),
+            error: Some(error),
+        }
+    }
+
+    /// Data pipe for sending to and receiving from the forwarded port.
+    pub fn stream(&mut self) -> Option<impl AsyncRead + AsyncWrite + Unpin> {
+        self.stream.take()
```

Comment:
i guess it's because this is also replacing the inner stream with a `None` - but we could do that in another way.
at the very least we could do the unwrap in here (since it's guaranteed)

Reply:
actually, i guess it's up to the user to not call this more than once - and that's why we probably should not do the unwrap.